tiled/database/orm.py

```diff
@@ -0,0 +1,195 @@
+        if value is not None:
+            if not isinstance(value, uuid_module.UUID):
+                raise ValueError(f"Expected uuid.UUID, got {type(value)}")
+            return value.bytes
+
+    def process_result_value(self, value, dialect):
+        if value is not None:
+            return uuid_module.UUID(bytes=value)
+
+
+class Timestamped:
+    """
+    Mixin for providing timestamps of creation and update time.
+
+    These are not used by application code, but they may be useful for
+    forensics.
+    """
+
+    time_created = Column(DateTime(timezone=True), server_default=func.now())
+    time_updated = Column(
+        DateTime(timezone=True), onupdate=func.now()
+    )  # null until first update
+
+    def __repr__(self):
+        return (
+            f"{type(self).__name__}("
+            + ", ".join(
+                f"{key}={value!r}"
+                for key, value in self.__dict__.items()
+                if not key.startswith("_")
+            )
+            + ")"
+        )
+
+
+principal_role_association_table = Table(
+    "principal_role_association",
+    Base.metadata,
+    Column("principal_id", Integer, ForeignKey("principals.id"), primary_key=True),
+    Column("role_id", Integer, ForeignKey("roles.id"), primary_key=True),
+)
+
+
+class Principal(Timestamped, Base):
+    __tablename__ = "principals"
+
+    # This id is internal, never exposed to the user.
+    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
+    # This uuid is public.
+    # SQLite does not support UUID4 type, so we use generic binary.
+    uuid = Column(
+        UUID,
+        index=True,
+        nullable=False,
+        default=lambda: uuid_module.uuid4(),
+    )
+    type = Column(Enum(PrincipalType), nullable=False)
+    display_name = Column(Unicode(255), nullable=False)
+    # In the future we may add other information.
+
+    identities = relationship("Identity", back_populates="principal")
+    api_keys = relationship("APIKey", back_populates="principal")
+    roles = relationship(
+        "Role", secondary=principal_role_association_table, back_populates="principals"
+    )
+    sessions = relationship("Session", back_populates="principal")
+
+
+class Identity(Timestamped, Base):
+    __tablename__ = "identities"
+
+    # An (id, provider) pair must be unique.
+    id = Column(Unicode(255), primary_key=True, nullable=False)
+    provider = Column(Unicode(255), primary_key=True, nullable=False)
+    principal_id = Column(Integer, ForeignKey("principals.id"), nullable=False)
+    # In the future we may add a notion of "primary" identity.
+
+    principal = relationship("Principal", back_populates="identities")
+
+
+class Role(Timestamped, Base):
+    __tablename__ = "roles"
+
+    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
+    name = Column(Unicode(255), index=True, unique=True)
+    description = Column(Unicode(1023), nullable=True)
+    scopes = Column(JSONList, nullable=False)
```

Comment:
JSONlist and not JSONList(255)?

Reply:
Good catch!

Reply:
Actually, now that I think about it, I left this intentionally unbounded, following JupyterHub's example. The list of scopes may be long.

Reply:
If the list of scopes is long and approaching unbounded, then should we put them in a mapping table rather than stuffing them in an unbounded column?

Reply:
Fair point. Given the choice I prefer putting a bound on the list.

Reply:
(For now)